schema_mutations, migration_manager: Ignore empty partitions in per-table digest

tgrabiec opened the pull request:
Schema digest is calculated by querying for mutations of all schema
tables, then compacting them so that all tombstones in them are
dropped. However, even if the mutation becomes empty after compaction,
we still feed its partition key. If the same mutations were compacted
prior to the query, because the tombstones expire, we won't get any
mutation at all and won't feed the partition key. So schema digest
will change once an empty partition of some schema table is compacted
away.

Tombstones expire 7 days after schema change which introduces them. If
one of the nodes is restarted after that, it will compute a different
table schema digest on boot. This may cause performance problems. When
sending a request from coordinator to replica, the replica needs
schema_ptr of exact schema version request by the coordinator. If it
doesn't know that version, it will request it from the coordinator and
perform a full schema merge. This adds latency to every such request.
Schema versions which are not referenced are currently kept in cache
for only 1 second, so if request flow has low-enough rate, this
situation results in perpetual schema pulls.

After ae8d2a550d227de2221c310bd38704ad2e078b35 (5.2.0), it is more liekly to
run into this situation, because table creation generates tombstones
for all schema tables relevant to the table, even the ones which
will be otherwise empty for the new table (e.g. computed_columns).

This change inroduces a cluster feature which when enabled will change
digest calculation to be insensitive to expiry by ignoring empty
partitions in digest calculation. When the feature is enabled,
schema_ptrs are reloaded so that the window of discrepancy during
transition is short and no rolling restart is required.

A similar problem was fixed for per-node digest calculation in
c2ba94dc39e4add9db213751295fb17b95e6b962. Per-table digest calculation
was not fixed at that time because we didn't persist enabled features
and they were not enabled early-enough on boot for us to depend on
them in digest calculation. Now they are enabled before non-system
tables are loaded so digest calculation can rely on cluster features.

Fixes #4485.

Manually tested using ccm on cluster upgrade scenarios and node restarts.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2144/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2144/

https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2144/artifact/testlog/x86_64/debug/cql-pytest.test_tools.1.log
```
db/schema_tables.cc:3133:47: runtime error: member call on null pointer of type 'replica::database'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior db/schema_tables.cc:3133:47 in 
Aborting on shard 0.
Backtrace:
  0xd4f4350
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x48197bc
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x4819413
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x466bae9
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x469ae85
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x478c669
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x478c89e
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x478c6ea
  /lib64/libc.so.6+0x3cb1f
  /lib64/libc.so.6+0x8ce5b
  /lib64/libc.so.6+0x3ca75
  /lib64/libc.so.6+0x267fb
  0xd566eee
  0xd5649c0
  0xd57d44e
  0xd580da8
  0xd581ebe
  0x1023d275
  0x11f26151
  0x11f2345a
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x435851d
  /jenkins/workspace/scylla-master/scylla-ci/scylla/build/debug/seastar/libseastar.so+0x4b5f2a7
```

@tgrabiec says that the schema context is probably not initialized.

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2210/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2212/

tgrabiec commented:
In V2:
  - Fixed SIGSEGV in schema loader due to uninitialized db in schema_ctx
  - Fixed schema_change_test failure due to change in digest calculation. Updated expected digests.
  - Extended schema_change_test to verify that digest doesn't change after gc_clock is advanced past tombstone expiration point

kbr-scylla commented:
@tgrabiec who should review this PR?

avikivity commented:
How likely is this to happen? If likely, we should fall back to the old computation on mixed clusters To avoid unreconcilable schema while the cluster is being upgraded.

tgrabiec commented:
> How likely is this to happen? If likely, we should fall back to the old computation on mixed clusters To avoid unreconcilable schema while the cluster is being upgraded.

Since 5.2.0, it is guaranteed to happen after 7 days when you restart a single server but not the others.

We use old computation during rolling upgrade, this is controlled by a cluster feature. When the feature transitions to being enabled and it wasn't enabled at boot time, digests are recomputed. Normally that should happen on all nodes at more or less the same time.

tgrabiec commented:
On older versions, it can still happen, but scenarios to trigger are more complex.

avikivity commented on db/schema_tables.cc:
```diff
@@ -151,7 +151,8 @@ static future<> merge_tables_and_views(distributed<service::storage_proxy>& prox
     std::map<table_id, schema_mutations>&& views_before,
-    std::map<table_id, schema_mutations>&& views_after);
+    std::map<table_id, schema_mutations>&& views_after,
+    bool reload);
```
I don't understand the motivation. How will schemas change? Isn't it a no-op?

avikivity commented on service/migration_manager.cc:
```diff
@@ -109,6 +109,14 @@ void migration_manager::init_messaging_service()
+        if (!_feat.table_digest_insensitive_to_expiry) {
+            _feature_listeners.push_back(_feat.table_digest_insensitive_to_expiry.when_enabled([this] {
+                (void) with_gate(_background_tasks, [this] {
+                    return reload_schema();
```
So is the goal to reload the mutations just to re-run them through the digester? If we generated the mutations from the schema object, could be have avoided it (don't remember if we have such an ability).

tgrabiec replied:
We could generate digests from schema_ptr in memory. The reason I go through the schema merge path is to make sure that this process is synchronized with other schema changes which could be happening concurrently, and that I reuse the existing mechanism for redistributing schema_ptrs. It's less efficient but on the safe side. Having a single path means less risk due to the two paths diverging over time.

avikivity commented on test/boost/schema_change_test.cc:
```diff
@@ -831,14 +831,14 @@ SEASTAR_TEST_CASE(test_schema_digest_does_not_change) {
+        utils::UUID("75550bef-3a95-3901-be80-4540f7d8a311"),
+        utils::UUID("aff80a2a-4a72-35bb-9ac3-f851013610d0"),
+        utils::UUID("030100b2-27aa-32f2-8964-0090a1af75f8"),
+        utils::UUID("16ba4b2d-7c61-393b-ba51-0890e25f4e22"),
```
Let's keep the old ones too (test with/without).

avikivity commented:
Looks good. I didn't understand everything so I asked questions.

tgrabiec commented:
In V2:
  - preserve test cases which check the old digests without the new feature enabled, in addition to the new digest.
  - drop the last patch which verified that expiry doesn't change table digests, which does not really add any value because table digests are taken from memory.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2452/

DoronArazii commented:
@nyh ping review please

tgrabiec commented:
ping @avikivity. You comments were addressed.

mykaul commented:
next-5.2 (https://jenkins.scylladb.com/job/scylla-5.2/job/next/334/consoleFull#874834529dae97f28-980f-4dce-a662-1b9b15a76cc1 ) seem to fail:
```
 service/migration_manager.cc:99:28: error: use of undeclared identifier 'reload_schema'
00:17:04                      return reload_schema();
```
